File: Assets/StreamingAssets/LUA/Quest.lua
-------------------------------------------------------
-- Project Porcupine Copyright(C) 2016 Team Porcupine
-- This program comes with ABSOLUTELY NO WARRANTY; This is free software,
-- and you are welcome to redistribute it under certain conditions; See
-- file LICENSE, which is part of this source code package, for details.
-------------------------------------------------------

-- HOWTO Log:
--ModUtils.ULog("Testing ModUtils.ULogChannel")
--ModUtils.ULogWarning("Testing ModUtils.ULogWarningChannel")
--ModUtils.ULogError("Testing ModUtils.ULogErrorChannel") -- Note: pauses the game

---------------------------- Quests Actions --------------------------------

function Quest_Have_Furniture_Built(goal)
    goal.IsCompleted = false
    local type = goal.Parameters["type"].Value
    local amount = goal.Parameters["amount"].ToInt()
    local amountFound = GameController.CurrentWorld.FurnitureManager.CountWithType(type)
    if (amountFound >= amount) then
        goal.IsCompleted = true
    end
end

function Quest_Spawn_Inventory(reward)
    local type = reward.Parameters["type"].Value
    local amount = reward.Parameters["amount"].ToInt()
    local inv = Inventory.__new(type, amount, amount)
    local tile = GameController.CurrentWorld.InventoryManager.GetFirstTileWithValidInventoryPlacement(6, GameController.CurrentWorld.GetCenterTile(), inv)
    if (tile == nil) then
        return
    end
    GameController.CurrentWorld.InventoryManager.PlaceInventory(tile, inv)
    reward.IsCollected = true;
end

ModUtils.ULog("Quest.lua loaded")
return "LUA Script Parsed!"

File: Assets/Scripts/Models/Quest/Quest.cs
#region License
// ====================================================
// Project Porcupine Copyright(C) 2016 Team Porcupine
// This program comes with ABSOLUTELY NO WARRANTY; This is free software,
// and you are welcome to redistribute it under certain conditions; See
// file LICENSE, which is part of this source code package, for details.
// ====================================================
#endregion

using System.Collections.Generic;
using MoonSharp.Interpreter;
using Newtonsoft.Json.Linq;

[MoonSharpUserData]
public class Quest : IPrototypable
{
    public string Name { get; set; }

    public string Type
    {
        get { return Name; }
    }

    public string Description { get; set; }

    public List<QuestGoal> Goals { get; set; }

    public bool IsAccepted { get; set; }

    public bool IsCompleted { get; set; }

    public List<QuestReward> Rewards { get; set; }

    public List<string> RequiredQuests { get; set; }

    /// <summary>
    /// Reads the prototype from the specified JObject.
    /// </summary>
    /// <param name="jsonProto">The JProperty containing the prototype.</param>
    public void ReadJsonPrototype(JProperty jsonProto)
    {
        Name = jsonProto.Name;
        JToken innerJson = jsonProto.Value;

        Goals = new List<QuestGoal>();
        Rewards = new List<QuestReward>();
        RequiredQuests = new List<string>();

        Description = PrototypeReader.ReadJson(Description, innerJson["Description"]);

        if (innerJson["RequiredQuests"] != null)
        {
            foreach (JToken token in innerJson["RequiredQuests"])
            {
                RequiredQuests.Add((string)token);
            }
        }

        if (innerJson["Goals"] != null)
        {
            foreach (JToken token in innerJson["Goals"])
            {
                QuestGoal goal = new QuestGoal();
                goal.ReadJsonPrototype(token);
                Goals.Add(goal);
            }
        }

        if (innerJson["Rewards"] != null)
        {
            foreach (JToken token in innerJson["Rewards"])
            {
                QuestReward reward = new QuestReward();
                reward.ReadJsonPrototype(token);
                Rewards.Add(reward);
            }
        }
    }
}